pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -0,0 +1,263 @@
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=pods/status,verbs=get;patch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/finalizers,verbs=update
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=resourceflavors,verbs=get;list;watch
+
+var NewReconciler = jobframework.NewGenericReconciler(
+	func() jobframework.GenericJob {
+		return &Pod{}
+	}, nil)
+
+type Pod corev1.Pod
+
+var _ jobframework.GenericJob = (*Pod)(nil)
+var _ jobframework.JobWithCustomStop = (*Pod)(nil)
+
+// Object returns the job instance.
+func (p *Pod) Object() client.Object {
+	return (*corev1.Pod)(p)
+}
+
+func (p *Pod) gateIndex() int {
+
+	for i := range p.Spec.SchedulingGates {
+		if p.Spec.SchedulingGates[i].Name == SchedulingGateName {
+			return i
+		}
+	}
+	return gateNotFound
+}
+
+// IsSuspended returns whether the job is suspended or not.
+func (p *Pod) IsSuspended() bool {
+	return p.gateIndex() != gateNotFound
+}
+
+// Suspend will suspend the job.
+func (p *Pod) Suspend() {
+	// TODO: maybe change the framework so this can provide feedback,
+	// the pod can only be "suspended" by the mutation hook if it's not
+	// done the only way to potentialy stop its execution is the eviction
+	// which will also terminate the pod.
+}
+
+// RunWithPodSetsInfo will inject the node affinity and podSet counts extracting from workload to job and unsuspend it.
+func (p *Pod) RunWithPodSetsInfo(podSetsInfo []jobframework.PodSetInfo) error {
+	if len(podSetsInfo) != 1 {
+		return fmt.Errorf("%w: expecting 1 got %d", jobframework.ErrInvalidPodsetInfo, len(podSetsInfo))
+	}
+	idx := p.gateIndex()
+	if idx != gateNotFound {
+		p.Spec.SchedulingGates = append(p.Spec.SchedulingGates[:idx], p.Spec.SchedulingGates[idx+1:]...)
+	}
+
+	// TODO: manage the node selector
+	// NOTE: it's only possible to add and only if k8s > 1.27 is used, case in which, since if the provided
+	// selectors are changing a existing key will fail we should be able to "refuse" the assignment
+
+	// if k8s < 1.27 TODO: wait for Version check patch
+	info := podSetsInfo[0]
+	if false {
+		if len(info.NodeSelector) > 0 {
+			return fmt.Errorf("%w: node selectors cannot be changed in k8s < 1.27", jobframework.ErrInvalidPodsetInfo)
+		}
+	} else {
+		ns := p.Spec.NodeSelector
+		if len(ns) > 0 {
+			overrideNS := make([]string, 0, len(ns))
+			for k, val := range ns {
+				if newVal, found := info.NodeSelector[k]; found && newVal != val {
+					overrideNS = append(overrideNS, k)
+				}
+			}
+			if len(overrideNS) > 0 {
+				return fmt.Errorf("%w: node selectors %s cannot be changed", jobframework.ErrInvalidPodsetInfo, strings.Join(overrideNS, ","))
```

Comment:
Isn't there a possibility that an existing node selector will be updated? 

The test in `job_controller_test.go` seems to be validating such behavior in the Job reconciler: https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobs/job/job_controller_test.go#L200-L224

Reply:
Thanks for flagging. I think that test is actually unnecessary.

The scheduler would make sure that such scenario doesn't happen:
https://kueue.sigs.k8s.io/docs/concepts/resource_flavor/#resourceflavor-labels
https://github.com/kubernetes-sigs/kueue/blob/6d30b82d1459590cdc553d9eb8f88f553635500b/pkg/scheduler/flavorassigner/flavorassigner.go#L358-L365

An admin user could actually try to manually modify a Workload object leading to  violating the check you are trying to make.
But we can just let the apiserver fail it, instead of Kueue.